lib/bridge/IrcBridge.js

```diff
@@ -754,6 +769,128 @@ IrcBridge.prototype.getBotClient = function(server) {
     });
 }
 
+
+var _solveTimeout = null;
+
+IrcBridge.prototype._updateVisibilityMap = function(isMode, key, value) {
+    let hasChanged = false;
+    if (isMode) {
+        if (typeof value !== 'boolean') {
+            throw new Error('+s state must be indicated with a boolean');
+        }
+        if (this._visibilityMap.channelModes[key] !== value) {
+            this._visibilityMap.channelModes[key] = value;
+            hasChanged = true;
+        }
+    }
+    else {
+        if (typeof value !== 'string' || (value !== "private" && value !== "public")) {
+            throw new Error('Room visibility must = "private" | "public"');
+        }
+
+        if (this._visibilityMap.roomVisibilities[key] !== value) {
+            this._visibilityMap.roomVisibilities[key] = value;
+            hasChanged = true;
+        }
+    }
+
+    if (hasChanged) {
+        clearTimeout(_solveTimeout);
+
+        _solveTimeout = setTimeout(() => {
+            this._solveVisibility()
+        }, 10000);
+        return Promise.resolve();
+    }
+
+    return Promise.resolve();
+}
+
+IrcBridge.prototype._solveVisibility = Promise.coroutine(function*() {
+    // For each room, do a big OR on all of the channels that are linked in any way
+
+    this._visibilityMap.mappings = yield this.getStore().getAllMappings();
+
+    let roomIds = Object.keys(this._visibilityMap.roomVisibilities);
+
+    // level is the current depth of recursion
+    let shouldBePrivate = (roomId, checkedChannels) => {
+        // If any channel connected to this room is +s, stop early and call it private
+
+        // List first connected
+        let channels = this._visibilityMap.mappings[roomId];
+        //      = ['localhost #channel1', 'localhost #channel2', ... ]
+
+        // Filter out already checked channels
+        channels = channels.filter((c) => checkedChannels.indexOf(c) === -1);
+
+        // No channels mapped to this roomId
+        if (!channels) {
+            return false;
+        }
+
+        let anyAreSecret = channels.any((channel) => {
+            let channelIsSecret = this._visibilityMap.channelModes[channel];
+
+            // If a channel mode is unknown, assume it is secret
+            if (typeof channelIsSecret === 'undefined') {
```

Comment:
This is an interesting one: should we be doing this? It's the safe bet, but most likely this will be `undefined` on startup as the modes slowly trickle in. Should we be bailing from modifying the visibility at all if this is not set?